sw/device/lib/dif/dif_otbn.h

```diff
@@ -0,0 +1,171 @@
+#include <stddef.h>
+#include <stdint.h>
+
+// Header Extern Guard (so header can be used from C and C++)
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+/**
+ * Width of wide instructions in OTBN, in bit.
+ */
+#define OTBN_WLEN 256
+
+/**
+ * Configuration for initializing an OTBN device.
+ */
+typedef struct dif_otbn_config {
+  mmio_region_t base_addr;
+} dif_otbn_config_t;
+
+/**
+ * Internal state of a OTBN device.
+ *
+ * Instances of this struct must be initialized by `dif_otbn_init()` before
+ * being passed to other functions.
+ */
+typedef struct dif_otbn {
+  mmio_region_t base_addr;
+} dif_otbn_t;
+
+/**
+ * Generic return codes for the functions in this library.
+ */
+typedef enum dif_otbn_result {
+  /**
+   * The function succeeded.
+   */
+  kDifOtbnOk,
+  /**
+   * The function failed a non-specific assertion.
+   *
+   * This error is not recoverable.
+   */
+  kDifOtbnError,
+  /**
+   * There is a problem with the argument(s) to the function.
+   *
+   * This return code is only returned before the function has any side effects.
+   *
+   * This error is recoverable and the operation can be retried after correcting
+   * the problem with the argument(s).
+   */
+  kDifOtbnBadArg,
+
+} dif_otbn_result_t;
+
+/**
+ * Initialize a OTBN device using `config` and return its internal state.
+ *
+ * A particular OTBN device must first be initialized by this function
+ * before calling other functions of this library.
+ *
+ * @param config Configuration for initializing an OTBN device.
+ * @param otbn OTBN instance that will store the internal state of the
+ *             initialized OTBN device.
+ * @return `kDifOtbnBadArg` if `config` or `otbn` is `NULL`,
+ *         `kDifOtbnOk` otherwise.
+ */
+dif_otbn_result_t dif_otbn_init(const dif_otbn_config_t *config,
+                                dif_otbn_t *otbn);
+
+/**
+ * Reset OTBN device.
+ *
+ * Resets the given OTBN device by setting its configuration registers to
+ * reset values. Disables interrupts, output, and input filter.
+ *
+ * @param otbn OTBN instance
+ * @return `kDifOtbnBadArg` if `otbn` is `NULL`, `kDifOtbnOk` otherwise.
+ */
+dif_otbn_result_t dif_otbn_reset(const dif_otbn_t *otbn);
+
+/**
+ * Start the execution of the application loaded into OTBN.
+ *
+ * @param otbn OTBN instance
+ * @return `kDifOtbnBadArg` if `otbn` is `NULL`, `kDifOtbnOk` otherwise.
+ */
+dif_otbn_result_t dif_otbn_start(const dif_otbn_t *otbn);
+
+/**
+ * Is OTBN busy executing an application?
+ *
+ * @param otbn OTBN instance
+ * @param[out] busy OTBN is busy
+ * @return `kDifOtbnBadArg` if `otbn` is `NULL`, `kDifOtbnOk` otherwise.
+ */
+dif_otbn_result_t dif_otbn_is_busy(const dif_otbn_t *otbn, bool *busy);
+
+// TODO: Think about making src/len in the imem/dmem write/read functions words
+// instead of bytes. Easy for imem with 32b words, harder for dmem with WLEN
+// words. And ideally, we have them consistent: either both are words, or none
+// of them are words.
```

Comment:
I know other DIFs were pushed to use something like uint32_t if only word writes are supported. That would be possible for the IMEM size of OTBN, but the DMEM side would need uint256_t. Something to be figured out.

Reply:
We think these types should reflect the alignment required on the ibex side - as ibex can only do a 32bit load/store at a time, these should be `uint32_t*` if you want them to be at least word-aligned. We're not sure that the ibex-side pointers should be more aligned than 32-bit, as that seems to waste space when the ibex-size accesses are all 32-bits wide maximum. This is especially the case as we think the OTBN peripheral doesn't have the ability to take a single 256-bit write.

By all means document and enforce stricter alignment requirements on the `offset` arguments.